docs/proposals/clusterconfig.md

```diff
@@ -0,0 +1,110 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Proposes a mechanism for specifying cluster-wide runtime config
+for all kube system components via a single config file. This will
+provide an easy way to specify init settings that are shared between
+system components.
+
+## Motivation
+
+Primary motivation is enabling/disabling features that are not tied to
+an API group. API groups can be selectively enabled/disabled in the
+apiserver via the `--runtime-config` commandline flag, but there is
+currently no mechanism to toggle alpha features that are controlled by
+e.g. annotations. This means the burden of controlling whether such
+features are enabled in a particular cluster is on feature implementors;
+they must either define some ad hoc mechanism for toggling (e.g. flag
+on component binary) or else toggle the feature on/off at compile time.
+
+Providing a mechanism for specifying cluster-wide config makes
+deploying non-API features simpler for implementors and cluster admins.
+
+## Design
+
+Cluster-wide runtime config will be specified in a single yaml file
+loaded from a known, default path. TBD whether default config path
+should be overwriteable via flag/env.
+
+Strawman:
+```shell
+$ cat /etc/srv/kubernetes/cluster_config.yaml
+runtimeConfig:
+  enableAppArmor: true
+  enableFooBarBaz: false
```

Comment:
I don't follow the "two config maps per process type" part of the question.
Which config maps are you referring to here?


One config map for runtime enablement (to deliver the value into pod or
node), and one for the component config.  VS always reading your alpha
enablement out of your component config.

I bring this up because I haven't seen a lot of cross cutting enablement,
and it was usually between controller and node, or admission controller and
controller.  So my suspicion was that we'd have more paired changes (set
config A and B, restart scheduler then nodes).

Having a shared config map for runtime config seems like it introduces more
coupling at runtime (a rolling update has to update the shared config and
the component configs, vs just updating each component config as you go).
Not a huge problem, but it's also one more thing you have to track and
check when doing updates.

I like having one config per component as long as we possibly can - I don't
think it breaks our model to have two sources of config, but there are some
advantages to duplicating a small number of sparse fields.